Support for drag and resize operations

RobbinMarcus opened the pull request:
This is a cleaned up version that merges both changes for the drag operation and resize operation. I implemented the requested changes by @jadahl and @linkmauve and solved conflicts generated by the old pull requests.

Merged these two pull requests together:
- Add support for drag operation on borderless window #987
- Support for resize operation on borderless windows #991

Affected issues:
- Provide a way to specify the window drag area on an undecorrated window #923
- Support for resizing undecorated window #990

Current state:
I tested these changes on windows 10 (win32). I changed the "window" example locally to have an undecorated window and added a buttons to test the drag and resize operations (glfwDragWindow or glfwResizeWindow).

Specifically for win32 this PR implements the `WM_NCHITTEST` and `WM_NCCALCSIZE` to get rid of the 7px white header above the window. These changes led me to add 2 functions to customize the caption area and the resize border size. Currently this information is available for every platform, but only used for win32. 

![UndecoratedResize](https://user-images.githubusercontent.com/4852449/192476799-cca65ae3-ed2d-48ef-a56d-b24d306a6661.gif)

Reviewers: please let me know if this needs to be split up into separate PRs. The drag/resize operations are mostly separate from the win32 specific undecorated window changes. 

Testing: This PR is only tested in windows 10 at the moment. The other platforms are in the same state as the previous PRs mentioned above, except for the changes I made to get them up to date to the latest master branch.

RobbinMarcus committed:
Support for drag and resize operations

RobbinMarcus committed:
Remove trailing whitespace (Travis CI build)

RobbinMarcus committed:
WL rename to serial (Travis CI build)

RobbinMarcus committed:
Merge remote-tracking branch 'upstream/master'

xteo commented:
@RobbinMarcus : Thanks for this MR
I am trying to setup our Application ( using GLFW) to be without OS Title bar so we can control the Look of the Titlebar fully like other moderm application 

I will try you MR this week but wanted to ask a question. in your note you mentioned you still have issue with Resize ? did I understand well ? 

** Undecorated glfwResizeWindow does not work, I cannot figure out why.**

Thanks for your contribution hope to get it working!

RobbinMarcus commented:
Hi @xteo. 
Yes that's correct, trying to resize the window does not work in this PR. 
I've tried to fiddle with some things but I couldn't get it to work. Good luck!

rokups commented:
This feature would be very helpful to Dear ImGui, so i was testing this functionality a bit. X11 implementation has an issue where drag operation makes mouse button stuck in pressed state even after drag operation is done and mouse button is released (which happens over window by the way). If anyone has any ideas how to work this around - please speak 🙏🏻

emoon commented:
It would be great to get this feature in. It's something I want myself as well

rokups commented:
I spent some time ironing it out. It works quite good on X11 now, but in the process i had to also fix inconsistent mouse button reporting. On windows however resizing seems to not work at all. I am not sure why, but maybe application has to support handling these resize messages..?

JasonTheKitten commented:
I believe this PR would be extremely useful.
Many times, one wishes to use an undecorated window. Perhaps they want their custom window decorations to have a different aesthetic. Or perhaps they want to add extra widgets to the decoration.
While it is possible (and easy) to manually write code to drag an undecorated window, it may not encompass all of the features of the default drag behavior of the OS's windowing system.
For example, on Windows, the default windowing system will allow you to maximize windows if you drag them to the top, or do a side-by-side view if you drag them to the sides.

This is why I would like this PR to be merged.

Cons-Cat commented:
I would certainly use this!

CybernetHacker14 commented:
For the issue - "Undecorated glfwResizeWindow does not work, I cannot figure out why" - can this solution work - https://github.com/melak47/BorderlessWindow ? I have tested it out, and it seems to be exactly equivalent to what the resultant output should turn out to be.

RobbinMarcus committed:
Merge remote-tracking branch 'upstream/master'

RobbinMarcus committed:
Add WS_THICKFRAME for resizing undecorated windows

RobbinMarcus commented:
> For the issue - "Undecorated glfwResizeWindow does not work, I cannot figure out why" - can this solution work - https://github.com/melak47/BorderlessWindow ? I have tested it out, and it seems to be exactly equivalent to what the resultant output should turn out to be.

I've added `WS_THICKFRAME` to the window style which seems to do the trick to enable window resizing while undecorated.

There's a 'beauty problem' where you will get the following result with this small white frame:
![Ryne_589utCQFhf](https://user-images.githubusercontent.com/4852449/192099782-748bbe7b-c912-41ab-b41d-64066eab733c.png)

It seems this is expected behavior though: https://stackoverflow.com/questions/39731497/create-window-without-titlebar-with-resizable-border-and-without-bogus-6px-whit
Removing the top bar requires additional work to fix and be compatible with all windows versions.

RobbinMarcus committed:
Attempt to fix build errors

Removed some changes that slipped past me during the latest merge.

CybernetHacker14 commented:
The problem now is that, with the `WS_THICKFRAME` option, dragging the window seems to not work, cause when the cursor is near the border, the resize prompt is thrown, instead of dragging. I am testing this on Windows, comparing it with the example repo which I have posted in my earlier comment.

RobbinMarcus committed:
Win32: set mouse button released after drag/resize

When calling the drag/resize operation GLFW would keep the left mouse button pressed even after drag/resize was completed. The other changes will hopefully make the build compile for other platforms.

RobbinMarcus committed:
Update to platform function setup

CybernetHacker14 commented:
- This is my specific functionality just for the windows - https://github.com/CybernetHacker14/glfw/tree/win32-borderless . I have 
implemented the `WM_NCHITTEST` and `WM_NCCALCSIZE` functionality.
- This is extremely platform specific, with code taken up from sections of - https://github.com/melak47/BorderlessWindow .

RobbinMarcus committed:
Undecorated border fix + customizable caption/bordersize

Added functions to set caption area and resize border size for undecorated windows. Implemented this functionality for win32 using WM_NCHITTEST and WM_NCCALCSIZE.

RobbinMarcus commented:
Thanks for the input @CybernetHacker14. I took up a part of the code to implement the win32 `WM_NCHITTEST` and `WM_NCCALCSIZE`. The resulting window looks pretty good! 

I added customizability options to set a rectangle area for the caption (so you can have some menu buttons where the window won't be dragged) and the border size where you can start the window resize. 

The defaults for these are as follows:
- Caption Area will be offset by the border to start at (0, 0) in the client area. It's size will be full-width and 16px in height. 
- The resize border size is 4px wide for all borders.

I have also added `WS_MAXIMIZEBOX` back in the style which allows the window to be maximized when you drag it to the top of the screen for example.

In the example below I set the caption area to be the top right side, so pressing in the top left of the window won't start the drag operation. This is in addition to the already implemented callable drag/resize functions which still work the same.

![UndecoratedResize](https://user-images.githubusercontent.com/4852449/192385172-073f6556-dcc0-4888-aa36-68c9866c1047.gif)

RobbinMarcus committed:
Fix appveyor build

CybernetHacker14 commented:
Looks pretty good. Check whether you can double click to maximize and minimize the windows. Also, see how the window behaves when you drag-dock it to different sides of the screen.

CybernetHacker14 commented:
> Thanks for the input @CybernetHacker14. I took up a part of the code to implement the win32 `WM_NCHITTEST` and `WM_NCCALCSIZE`. The resulting window looks pretty good!
> 
> I added customizability options to set a rectangle area for the caption (so you can have some menu buttons where the window won't be dragged) and the border size where you can start the window resize.
> 
> The defaults for these are as follows:
> 
> * Caption Area will be offset by the border to start at (0, 0) in the client area. It's size will be full-width and 16px in height.
> * The resize border size is 4px wide for all borders.
> 
> I have also added `WS_MAXIMIZEBOX` back in the style which allows the window to be maximized when you drag it to the top of the screen for example.
> 
> In the example below I set the caption area to be the top right side, so pressing in the top left of the window won't start the drag operation. This is in addition to the already implemented callable drag/resize functions which still work the same.
> 
> ![UndecoratedResize](https://user-images.githubusercontent.com/4852449/192385172-073f6556-dcc0-4888-aa36-68c9866c1047.gif) [ ![UndecoratedResize](https://user-images.githubusercontent.com/4852449/192385172-073f6556-dcc0-4888-aa36-68c9866c1047.gif) ](https://user-images.githubusercontent.com/4852449/192385172-073f6556-dcc0-4888-aa36-68c9866c1047.gif) [ ](https://user-images.githubusercontent.com/4852449/192385172-073f6556-dcc0-4888-aa36-68c9866c1047.gif)

Does this exclude the area where the "Test no dragging" button is present, or there is some callback type thing, where you are checking if any button is hovered or not.

RobbinMarcus commented:
> Looks pretty good. Check whether you can double click to maximize and minimize the windows. Also, see how the window behaves when you drag-dock it to different sides of the screen.

All the normal windows 10 functionality now works similar to the decorated version as far as I can tell. 

> Does this exclude the area where the "Test no dragging" button is present, or there is some callback type thing, where you are checking if any button is hovered or not.

I'm explicitly not checking any window contents as this will be very application specific. Without the caption area I implemented the window would be dragged if you pressed on the "Test no drag" button there. The `WM_NCHITTEST` overrides any other application behavior by returning that it hit the caption.

Scr3amer commented:
Hello !

I wanted to play with your branch.
When I compile with clang 14 I have a warning:

win32_window.c:1896:3: warning: variable 'wBorder' is used uninitialized whenever switch default is taken

Should we early return ?
Or set wBorder to HTBORDER ?
Or any other option ?



PS:
I was trying to do exactly that but just using GLFW API but no luck. (It worked for the bottom and right side but resizing and repositioning the window at the same time was buggy for left and top)

RobbinMarcus committed:
Early return for clang compile

Changed functions names to be in line with other glfw functions

Scr3amer commented:
> I've added an early return for this, hope this works. If this default case hits, it will trigger the assert as this should not happen, but adding the return should make it compile for clang.

So we don't want to always call ReleaseCapture ? I am not very familiar with Win32 API (even though I read the doc of that function)

> I did also notice this, but I don't think there is a use case for continually swapping between resizable or not. I'll have to look into this though, probably have to shrink the window by the border size whenever we go back to undecorated.

I agree with you, it should be set only once. It was more for the polish than the functional aspect of it.

> Do you mean you were trying to replicate my example from above and couldn't resize or move? Did you use the new functions to increase the caption size and resize border:

Sorry I was not very clear in what I said. I meant initially I tried to implement the feature of this PR using only the GLFW API with glfwSetWindowSize followed by glfwSetWindowPos but it didn't work for left and top borders. So I went on Github and found this PR :). That's all.

Scr3amer commented:
Me again.

It works fine on my PC (Win10 + Clang14 UCRT MinGW)

I just noticed one last thing, when we resize ( or drag) it pauses the event loop meaning the content doesn't scale dynamically. It is paused  and then refreshed only when we release the mouse.

While when we resize Firefox, Chrome, VS Code etc.... The content is refreshed and updated while we resize / drag the window.

Scr3amer commented:
Hello hello !

Me again, just adding some things I noticed while I use it.

Overall It works like a charm on my PC (Win10 + Clang14 UCRT MinGW)

3 things I noticed though:

- the small detail of the size increase on toggle of the decoration I mentionned earlier
- why not setting GLFW_INVALID_ENUM for glfwResizeWindow if border is wrong ?
**-  when we drag the window to the top it maximizes but then dragging it down doesn't minimize it again. Feels like a missing feature, do you agree ?**

Cheers

AlfredXiaSF commented:
> Hello hello !
> 
> Me again, just adding some things I noticed while I use it.
> 
> Overall It works like a charm on my PC (Win10 + Clang14 UCRT MinGW)
> 
> 3 things I noticed though:
> 
> * the small detail of the size increase on toggle of the decoration I mentionned earlier
> * why not setting GLFW_INVALID_ENUM for glfwResizeWindow if border is wrong ?
> * **sometimes dragging the top border of a maximized window doesn't "register"** and the window is not minimized. Especially if the mouse cursor is at y = 0 but it can also happen within the border.
> 
> Cheers

Hello, @Scr3amer 
> sometimes dragging the top border of a maximized window doesn't "register" and the window is not minimized. Especially if the mouse cursor is at y = 0 but it can also happen within the border.
I try to make some changes in `HitTest`, seems fix this. The key is, if Maximized set `clientAreaTop` to zero.

```cpp
static int HitTest(_GLFWwindow* window, RECT windowRect, POINT cursor, POINT border)
{
    const int captionOffsetX = window->captionOffsetX != GLFW_DONT_CARE ? window->captionOffsetX : 0;
    const int captionOffsetY = window->captionOffsetY != GLFW_DONT_CARE ? window->captionOffsetY : 0;
    const int captionSizeX = window->captionSizeX != GLFW_DONT_CARE ? window->captionSizeX : windowRect.right - windowRect.left;
    const int captionSizeY = window->captionSizeY;

    const int clientAreaLeft = windowRect.left + border.x;
    const int clientAreaRight = windowRect.right - border.x;
    const int clientAreaTop = windowRect.top + (_glfw.platform.windowMaximized(window) ? 0 : border.y);
    const int clientAreaBottom = windowRect.bottom - border.y;

    const int cursorInCaption =
        cursor.x >= windowRect.left + captionOffsetX &&
        cursor.x <= windowRect.left + captionOffsetX + captionSizeX &&
        cursor.y >= windowRect.top + captionOffsetY &&
        cursor.y <= windowRect.top + captionOffsetY + captionSizeY;
```
---
 Case I only want to set CaptionArea height, so I set `offsetX` `offsetY` and `sizeX` to `GLFW_DONT_CARE`, but it not work, so I change `glfwSetWindowCaptionArea`
```cpp
GLFWAPI void glfwSetWindowCaptionArea(GLFWwindow* handle, int offsetX, int offsetY, int sizeX, int sizeY)
{
    _GLFWwindow* window = (_GLFWwindow*)handle;
    assert(window != NULL);

    _GLFW_REQUIRE_INIT();

    if (offsetX != GLFW_DONT_CARE && offsetX < 0)
        return;
    if (offsetY != GLFW_DONT_CARE && offsetY < 0)
        return;
    if (sizeX != GLFW_DONT_CARE && sizeX < 1)
        return;
    if (sizeY < 1)
        return;

    window->captionOffsetX = offsetX;
    window->captionOffsetY = offsetY;
    window->captionSizeX = sizeX;
    window->captionSizeY = sizeY;
}
```

sasichkamega commented:
why not merged?

nulldoubt commented:
When will this be merged? Anything holding us back? 🤔🤔

RobbinMarcus commented:
There has been no comment from reviewers on this pull request for 6 years now. 
From their perspective I could see maintenance would increase if they would support this option: if it doesn't work on any of the platforms they would have to fix it. 
It could be they have decided there is no place for this functionality in GLFW and/or the amount of people that would use it is too small (though I would appreciate a reply with that information so we can close it). 

I haven't taken the time to fix the conflicts for this PR with the latest release of GLFW, because I think it would be a waste of time if this issue would be in conflict again after a couple changes and then remain like this for another 6 years!

Give this message a thumbs up if you use or would use this functionality. I'll update this PR to resolve the conflicts with the latest if I get over 100 on this comment.

alexv-ds commented:
@RobbinMarcus ✊

constantitus commented:
I'm a bit confused.
Why does `resizeBorderSize` only affect the windows implementation ? I thought that since I have to manage borders myself on the other platforms, I'll just set it to zero to disable it, but I can't do that either ?
```c
GLFWAPI void glfwSetWindowResizeBorderSize(GLFWwindow* handle, int size)
{
...
    if (size < 1)
        return;
```
Same with the captionArea, it's only checked on windows and you can't fully disable it.
```c
GLFWAPI void glfwSetWindowCaptionArea(GLFWwindow* handle, int offsetX, int offsetY, int sizeX, int sizeY)
{
...
    if (offsetX < 0 || offsetY < 0 || sizeX < 1 || sizeY < 1)
        return;
```
Why ?

RobbinMarcus commented:
Back when I implemented this I only ran this on Windows, not taking into account any of the other platforms except the work done by previous PRs. 

I don't see much traction for this PR to be merged so I'm going to close it. Feel free to use any or all code from this PR if you want to split it up into separate parts or update it to the latest.
